usagov/usagov-2021
# Usagov 1170 nginx config

## Jira Task

https://cm-jira.usa.gov/browse/USAGOV-1170

## Description

I added rate limit to our cms and waf in our nginx config, to prevent HTTP/2 Rapid Reset Vulnerability.

## Type of Changes

  - Add screenshot showing what it should look like

## Testing Instructions


**Test CMS:**
1. Open a terminal and go to the project folder.
2. Run the these commands in the terminal:
    ```
    bin/init
    docker compose up
    ```
3. Go to Docker Desktop and open the cms terminal
<img width="1145" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/c0b3ba4d-1658-4634-8d86-2f6f7b280e71">

4. Type `cd web` in the cms terminal and hit enter.
5. Type `chmod u+x test_cms.sh` in the cms terminal and hit enter.
6. Type `./test_cms.sh` in the cms terminal and hit enter.
7. Go back to the terminal where you run `docker compose up` and use command (or Cmd) ⌘ + F to search the following text: "HTTP/1.1" 429".
8. As a result, you should see that approximately 150 requests had code 200 as a response and the remaining ones had code 429. _Note: the `test_cms.sh` file has a total of 160 requests._
Example: 
<img width="1230" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/c0557a0e-467f-4adf-b393-0df9c20ebe86">


**Test WAF:**
1. Go to CircleCI and make sure that this is the version in dev
<img width="1254" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/4e26a0ef-5bda-4479-92cd-2a3d4003452b">

3. Open two terminal windows and log in to cloud.gov.
4. Type `cf target -s dev` in the terminal and hit enter.
6. In one of the windows, type `cf ssh waf` in the terminal and hit enter.
7. Type `cd /etc` in the terminal and hit enter.
8. Type `chmod u+x test_waf.sh` in the terminal and hit enter.
9. Go to the other window and type 'cf logs waf`.
10. Go back to the windows where you ran `chmod u+x test_waf.sh`, type `./test_waf.sh` and hit enter.
Output:
<img width="476" alt="image" src="https://github.com/usagov/usagov-2021/assets/39020809/9ccdb546-8936-4b9e-9337-c6edde8cd91a">

11. Copy and paste that output in any tool you want and use command (or Cmd) ⌘ + F to search the following text: "HTTP/1.1" 429".
12. As a result, you should see that approximately 14 requests had code 429 as a response and the remaining ones had code 502, you can ignore that code. 

### Requires New Config

### Requires New Content

### Validation Steps
- Test instruction 1
- Test instruction 2
- Test instruction 3

## Security Review




## Reviewer Reminders
- Reviewed code changes
- Reviewed functionality
- Security review complete or not required

## Post PR Approval Instructions
Follow these steps as soon as you merge the new changes.

1. Go to the [USAGov Circle CI project](https://app.circleci.com/pipelines/github/usagov/usagov-2021).
13. Find the commit of this pull request.
14. Build and deploy the changes.
15. Update the Jira ticket by changing the ticket status to `Review in Test` and add a comment. State whether the change is already visible on [cms-dev.usa.gov](http://cms-dev.usa.gov/) and [beta-dev.usa.gov](http://beta-dev.usa.gov/), or if the deployment is still in process.
